xayn-ai/src/coi/relevance.rs

```diff
@@ -7,49 +7,91 @@ use derive_more::Into;
 
 use crate::coi::{key_phrase::KeyPhrase, CoiError, CoiId};
 
-/// A finite relevance score.
+/// A finite f32.
 #[derive(Clone, Copy, Debug, Default, Into, PartialEq, PartialOrd)]
-pub(crate) struct Relevance(f32);
+// invariant: the wrapped value must always be finite
+struct F32(f32);
+
+impl PartialEq<f32> for F32 {
+    fn eq(&self, other: &f32) -> bool {
+        self.0.eq(other)
+    }
+}
+
+impl Eq for F32 {
+    // never nan by invariant
+}
+
+#[allow(clippy::derive_ord_xor_partial_ord)]
+impl Ord for F32 {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap(/* never nan by invariant */)
+    }
+}
+
+/// A relevance score.
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub(crate) struct Relevance(Rel);
+
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum Rel {
```

Comment:
What about impl the `coi` and `kp` methods on the enum and removing the `Relevance` struct?

Reply:
in this case the issue is with enum visibility. since we would need to make it `pub(crate) enum Relevance` which means that the variants and especially the contained data is a public field too. in turn we would have to make `F32` public as well, otherwise we get the private-type-in-public-interface compiler error, but this is an internal implementational detail (in the end it's the whole purpose of the enum wrapping of coi/key phrase relevances that outside of this module noone has to deal with the internals, but if we were to expose it then people could match against the enum and split the fields apart out of context rendering the additional type safety void).

it would be nice to have public enums where variants can contain private associated data. i tried to work around that by wrapping a private enum in a public struct.